usr/src/uts/armv8/os/fakebop.c

```diff
@@ -287,6 +287,50 @@ boot_fb_shadow_init(void * ops __unused)
 	/* just a stub */
 }
 
+void
+bop_relocate(void)
+{
+	caddr_t vaddr;
+	extern caddr_t psm_map_phys(paddr_t addr, size_t len, int prot);
+	extern void boot_uart_relocate(caddr_t vaddr);
+
+	if (xbootp == NULL)
+		return;
+
+	if (xbootp->bi_fdt != 0) {
+		size_t sz;
+
+		if (fdt_check_header((const void *)xbootp->bi_fdt) != 0)
+			panic("provided FDT is invalid\n");
+
+		/*
+		 * This may be a little generous... we really should
+		 * not be editing the FDT any further.
+		 */
+		sz = fdt_totalsize((const void *)xbootp->bi_fdt);
```

Comment:
_we_ edit it pretty much to add a phandle to everything for finding nodes by id, we should really stop even doing that, for what it's worth.

I had wondered when I was looking at /dev/fdt whether we should relocate it into a _readonly_ mapping, but not yet.

Reply:
> add a phandle

This is done in inetboot now (and in the new loader implementation). It's probably worth tracking down where we're still required read/write access and mapping it as read-only (and without the extra space).

Reply:
Oh, I remembered where it happens. There is currently code in promif to create the expected top-level nodes when they don't exist (/chosen, /options etc.). The intetboot code does this _after_ fixing phandles, so accessing these nodes then trigger the phandle code in promif.

The `loader(7)` change reorders the mixups to address this , which leaves is with `prom_setprop`, which would be unused. Once we delete that we can move the mapping to read-only..

Reply:
I _think_ that u-boot should add these "missing" props for us now, too?

Reply:
I'm afraid not. Mostly people won't care too much about phandles unless there are cross-node references - we need them everywhere because it's our glue between DDI and the objects in the DTB.

Reply:
Oh, sorry.  I meant /options and /chosen, not our kind of weird phandle obsession.